FEATURE: New health check

mhsdesign opened the pull request:
see https://github.com/neos/neos-development-collection/issues/4243 for overview

Thanks to everyone who helped ;)

this is the flow part of it 

neos part: https://github.com/neos/neos-setup/pull/12


how to try it out?

use this branch `feature/new-health-check` for `neos/setup` and `neos/neos-setup`

requires https://github.com/neos/flow-development-collection/pull/3045

```
    "neos/setup": "dev-feature/new-health-check as dev-master",
    "neos/neos-setup": "dev-feature/new-health-check",
```

in case youre trying to use the new setup in a dev collection, it will fail:
> They both replace neos/cli-setup and thus cannot coexist.

because `neos/cli-setup` is not yet removed.

run
`./flow setup`

and navigate to 

`/setup` in your browser ;)

ideally try installing a fresh neos with the help of those tools.


Todo:
- [x] https://github.com/neos/setup/issues/60
- [x] cache invalidation bug: the db settings cache are not emptied so updating the db connection doesnt let the check pass
- [x] test on windows

kitsunet committed:
WIP: simplified setup

kitsunet committed:
WIP: Remove old resources

kitsunet committed:
Fix request handler

kitsunet committed:
Remove old configuration

mhsdesign committed:
FEATURE: Implement Healthcheck

mhsdesign committed:
LOl

kitsunet committed:
TASK: Set JSON content type for compiletime checks

mhsdesign committed:
TASK: Introduce runtime ente to ente check

kitsunet committed:
TASK: BasicRequirementsCheck

mhsdesign committed:
TASK: Apply suggestions from code review

mhsdesign committed:
TASK: Add missing supportedDatabaseDrivers

mhsdesign committed:
TASK: Move RequestHandler.php

mhsdesign committed:
TASK: Remove dead code

mhsdesign committed:
TASK: Add minimal js rendering

mhsdesign committed:
TASK: CompiletimeHealthcheckInterface

mhsdesign commented:
Todo:

- welcome rename
- welcome is currently only compiletime -> sub request with runtime?

mhsdesign committed:
TASK: `flow setup:welcome:index` run runtime checks

jonnitto committed:
TASK: Update some wordings

mhsdesign committed:
TASK: DoctrineHealthcheck

mhsdesign committed:
TASK: HealthCollection with keys

jonnitto committed:
TASK: Interface for setup

TODO: Build stack

kitsunet committed:
TASK: Refactor basic requirements check

Also adds check for php web/cli version mismatch

jonnitto committed:
TASK: Add shadow

mhsdesign committed:
TASK: Add buildstack and asset forwarding

mhsdesign committed:
TASK: Add jons shadow

mhsdesign committed:
Merge remote-tracking branch 'origin/feature/new-health-check' into feature/new-health-check

mhsdesign committed:
TASK

jonnitto committed:
TASK: Update Build stack

jonnitto committed:
TASK: Visual update

jonnitto committed:
TASK: Add watch task

jonnitto committed:
TASK: Adjust font handler

mhsdesign committed:
TASK: Fix `checkPhpBinaryVersion` and add types

jonnitto committed:
TASK: Stop after first check if failed

mhsdesign committed:
TASK: Refactor Endpoint

mhsdesign committed:
TASK: buildCompiletimeSequence

mhsdesign committed:
TASK: Fix html

jonnitto committed:
TASK: Move builded files to private

kitsunet committed:
TASK: Super minimal boot for compiletime checks and refactoring

jonnitto committed:
TASK: Set output commands to code

jonnitto committed:
TASK: Output html content on fatal error

mhsdesign committed:
TASK: Pre boot time cli command

mhsdesign committed:
TASK: Composer replace `neos/cli-setup`

jonnitto committed:
TASK: Visual update for YAML Keys

jonnitto committed:
TASK: Return HTML content if no valid JSON is returned

jonnitto committed:
TASK: Remove dependencies

kitsunet committed:
TASK: Prevent exception in migration check

Probably more Exception types could happen, but it's hard to
predict with doctrine, so for now these two should be enough.

mhsdesign committed:
TASK: Make js and css endpoints work on every server

mhsdesign committed:
TASK: Redirected Ambiguous Or Legacy CommandIdentifier

mhsdesign committed:
TASK: More sophisticated redirection of cli commands

now `flow help setup` will show the help for `setup:index`

mhsdesign committed:
TASK: Improve `./flow setup` cli output

mhsdesign committed:
TASK: Require php 8.1

mhsdesign committed:
TASK: If the health collection has an error, don't fetch the next endpoint.

mhsdesign committed:
TASK: Neos.Setup runtime route as first entry so we dont try to match against the NodeRoutePathHandler, which might throw because no HomePage exists

mhsdesign committed:
TASK: DatabaseHealthcheck assert that `dbname` is set, as otherwise a connection still can be established at first

... but we do need the database name ^^

mhsdesign committed:
TASK: Introduce HealthcheckEnvironment and make use of HealthcheckEnvironment::isSafeToLeakTechnicalDetails()

mhsdesign committed:
TASK: move construction of HealthcheckEnvironment outside of runner

mhsdesign commented on Classes/Infrastructure/Healthcheck/BasicRequirementsHealthcheck.php:
```diff
@@ -0,0 +1,287 @@
+        return 'Basic system requirements';
+    }
+
+    public function execute(HealthcheckEnvironment $environment): Health
+    {
+        $this->environment = $environment;
```
i dont like it, but whats the alternative? @kitsunet 

and a `setEnviroment` method on the interface (or better like in other places with is_callable) is also not the holy grail ...

kitsunet replied:
Yeah, wouldn't do. Either we introduce options you can set per check and set those options differnetly per environment, or we just disable all checks in production (again via settings) and people decide what they want to activate.

mhsdesign committed:
BUGFIX: ConfigurationManager without caching

requires https://github.com/neos/flow-development-collection/pull/3045

mhsdesign commented:
Why we run the first check pre-compile time:

- No errors will be thrown when there are invalid php sources (as we dont attempt to require many classes / might be because we trigger the reflection service?)
- On windows the compile boot might not succeed if the php version is not set correclly

mhsdesign committed:
BUGFIX: When compile time fails after it succeeded once remove outdated runtime check results

mhsdesign committed:
BUGFIX: Decide between few and many doctrine migrations to either warn or abort with error

This should prevent on windows:
doctrine migrate is that slow that the page will refresh and run into an error because doctrine is not yet migrated 😅 (missing table…) - find out if migration is pending

mhsdesign committed:
TASK: Add `isWindows` to execution environment

mhsdesign committed:
TASK: Allow to use `{{flowCommand}}` variable in messages

mhsdesign committed:
BUGFIX: Catch and log throwables when executing health check

mhsdesign committed:
TASK: Adjust composer dependencies and remove old migrations

mhsdesign committed:
BUGFIX: Place settings context aware in the folder

https://github.com/neos/setup/issues/60

mhsdesign committed:
TASK: Require php 8.1

mhsdesign committed:
TASK: Use correct FlowInvocationCommand in all places

mhsdesign committed:
WIP: Use subprocess to test php binary

mhsdesign committed:
TASK: Minor adjustments

mhsdesign committed:
TASK: Catch possible errors from runtime check subprocess

mhsdesign committed:
BUGFIX: Make checking for `checkPhpBinaryVersion` more relivable

On windows the cli will currently tell everything is green, even if flow cant build subprocesses.

`Scripts::buildPhpCommand` actually validates most of the things we already check. `Scripts::buildPhpCommand` only needs to be adjusted in flow to actually throw in case some check doesnt succeed.

| speed  | strategy                                                                                    |
|--------|---------------------------------------------------------------------------------------------|
| 0.026s | duplicating the php binary checks, using the legacy code from Neos.Setup                    |
| 0.065s | starting a dummy subprocess `Scripts::executeCommand` with minimal flow boot (instant exit) |
| 0.044s | throwing errors already in `Scripts::buildPhpCommand` if phpbinary not found (to be done)   |
| 0.025s | optimizing `Scripts::buildPhpCommand` - by reverting #2491 (this makes it twice as slow)    |

mhsdesign committed:
TASK: Remove dublicated checks which are enforced by the bootstrap and never run

mhsdesign committed:
TASK: Warn if user cannot create symlinks on windows

mhsdesign committed:
TASK: Improve exception messages

mhsdesign committed:
TASK: WIP: Native support for `oh-my-zsh-flow`? `flow` without `./`

mhsdesign committed:
TASK: Convert enums to php 8 pseudo enums

mhsdesign committed:
TASK: Remove `readonly` enum declaration for php 8.0 compat

mhsdesign committed:
TASK: Declare compatibility with php 8.0

mhsdesign committed:
BUGFIX: GUI, wait for answer before polling endpoint again

mhsdesign committed:
BUGFIX: Prevent flashing of "Message copied" when reloading

mhsdesign committed:
Revert "TASK: WIP: Native support for `oh-my-zsh-flow`? `flow` without `./`"

This reverts commit e298fab93e029b952cd876ab5a75b9c1eca6cdf4.

mhsdesign committed:
TASK: require upcoming flow patch release

mhsdesign committed:
TASK: Fix `writeSettings`

mhsdesign committed:
TASK: Add technical documentation

mhsdesign committed:
TASK: Fix style ci via cool patch
